libraries/chain/controller.cpp

```diff
@@ -2297,13 +2297,17 @@ struct controller_impl {
                                                        config::active_name}),
                          calculate_threshold( 2, 3 ) /* more than two-thirds */                      );
 
-      update_permission( authorization.get_permission({config::producers_account_name,
-                                                       config::majority_producers_permission_name}),
-                         calculate_threshold( 1, 2 ) /* more than one-half */                        );
+      const permission_object* major_perm = authorization.find_permission({config::producers_account_name,
+                                                                           config::majority_producers_permission_name});
+      if( major_perm ) {
+         update_permission( *major_perm, calculate_threshold( 1, 2 ) /* more than one-half */ );
+      }
```

Comment:
Yes, this is a hard-forking change and I should have mentioned that in the description. Obviously no chain has removed these permissions or they would have hit this issue. Your point is, of course, that once this is merged those with this change will continue a chain with the permission removed and those without the change will not. I hate to add a protocol feature for something like this. I guess another option would be to subjectively not allow removal of `prod.major` and `prod.minor`. Any other ideas?

Reply:
I think that subjectively disallowing the removal is the correct behavior.  This is assuming that my conclusion that a block removing the permissions cannot currently become irreversible is correct.  I'd like a second opinion confirming that before committing to a course of action.
- It is impossible to build on such a block because `start_block` always fails
- A block can only become irreversible after a descendant is validated

Therefore, I think the following semantics will preserve consensus safely:
- Reject blocks containing the removal
- Reject normal transactions containing the removal
- Subjectively reject deferred transaction containing the removal (cannot retire with objective fail)

This matches the semantics of subjective transaction failure.

Reply:
Oh, wait...

Removing and then recreating the permission is a thing.

- When producing a block subjectively reject transactions containing the removal
- When validating a block, reject the block if the permissions do not exist when the block is completed